### ISSUE TITLE
[HOLD for payment 2022-11-16] [$500] Draft messages jump around in the LHN when accessed a second time

### ISSUE THREAD

**Issue by kbecciv** [Expensify/App#11330]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Issue found when executing PR** https://github.com/Expensify/App/pull/11211

## Action Performed:
1. Go to staging.new.expensify.com
2. Login with the expensifail account
3. Make sure you are in default view mode
8. Add a draft to a chat
9. Verify its position in the LHN doesn't change
10. Select a different chat
11. Verify that the previous draft chat moves up to the top of the LHN to be grouped with other draft chats
12. Select the previous draft chat again

## Expected Result:
Draft chat stays at the top of the LHN

## Actual Result:
Draft chat jumps down to it's original position

## Workaround:
Unknown

## Platform:

Where is this issue occurring?

- Web
- iOS
- Android
- Desktop App
- Mobile Web

**Version Number:** 1.2.7.1

**Reproducible in staging?:** Yes

**Reproducible in production?:** Yes

**Email or phone of affected tester (no customers):** any

**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** Any additional supporting documentation
**Expensify/Expensify Issue URL:** 
**Issue reported by:** 
**Slack conversation:** 

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

https://user-images.githubusercontent.com/47436092/192620492-0cde8324-e2ae-42d6-b41c-b4401d41d9df.mov

**Comment by tgolen** [comment]
I just tested this on staging and cannot reproduce. You can see here how the draft reports are all listed alphabetically:

![image](https://user-images.githubusercontent.com/1228807/192617083-0b04080d-bb23-426d-8073-5f8e1e798b35.png)

**Comment by roryabraham** [comment]
I did some testing on mobile web and found that reports with draft comments appeared in the correct alphabetical order. However, this did not happen until I opened a different report and came back to the LHN, which is actually the same bug report as https://github.com/Expensify/App/issues/11332

I think that this:

> We purposefully don't change the order of the active chat when adding drafts (to keep the LHN from jumping around, which can be jarring to the user).

Makes sense on wide screens, but provides a UX that's not ideal on narrow screens (i.e: you come back to the LHN after writing a draft comment and the report is not pinned to the top in alphabetical order)

**Comment by tgolen** [comment]
Yes, maybe that's so, but to me, that's a very separate discussion.

**Comment by roryabraham** [comment]
I also notice that when you click on a report with a draft comment, it jumps down to it's previous unpinned position

**Comment by grgia** [comment]
I cannot reproduce the alphabetical order bug nor the drafted chat jumping down in staging. Are you still able to reproduce that one @roryabraham ? If not, I think I'll close this out.

**Comment by roryabraham** [comment]
Yes, I'm still able to reproduce this in `1.2.10-0`:

1. Select a chat with someone in the middle of your LHN
2. Draft a message
3. Click on another user in your LHN. The report with the draft message will jump to the top
4. Click back on the report with the draft message. It will jump back down to where is was before

**Comment by grgia** [comment]
I updated staging to `1.2.11-0` and I was able to reproduce the drafted message jumping. However, I'm not sure what the expected result would be here. Should it stick to the top until a message is sent or deleted (/when chat input is cleared)?

**Comment by tgolen** [comment]
I can also reproduce the draft message jumping back to its original position. I need to think about how this should be handled. I think we want:

1. When you add a draft, the chat stays in the same place in the LHN
2. When you go to another chat, the draft report should move up to the top of the LHN
3. When you go to the draft report again, it should stay at the top of the LHN

The logic in the code makes sense why it's jumping back to its original position (because if it's the active report, it won't be included in the group of drafted reports). So I need to consider how to make an exception for this UX.

The easiest solution is to change the ideal UX slightly so that the first point is:
1. When you add a draft, the chat jumps to the top of the LHN and then stays there

**Comment by tgolen** [comment]
@roryabraham Maybe that solution also solves your concern here:

> Makes sense on wide screens, but provides a UX that's not ideal on narrow screens (i.e: you come back to the LHN after writing a draft comment and the report is not pinned to the top in alphabetical order)

**Comment by grgia** [comment]
I would agree with that solution! In that case, I'm going to move this onward and apply the external label / unassign.

**Comment by mallenexpensify** [comment]
Posted https://www.upwork.com/jobs/~011c7454f55e6c9c1f
Thanks @tgolen for updating the OP.  I reassigned myself. I accidentally got unassigned. @grgia , when unassigning from an issue, please do so before applying labels to avoid accidental unassigning (this used to happen to me all the time, Sonia tipped me to why)

**Comment by deetergp** [comment]
We are still awaiting contributor proposals that that cause drafts to behave according to @tgolen's comment [here](https://github.com/Expensify/App/issues/11330#issuecomment-1265580009):

> The easiest solution is to change the ideal UX slightly so that the first point is:
> 
> 1. When you add a draft, the chat jumps to the top of the LHN and then stays there

**Comment by mallenexpensify** [comment]
Price Doubled to $500
https://www.upwork.com/jobs/~011c7454f55e6c9c1f

**Comment by tsa321** [comment]
### Proposal

https://github.com/Expensify/App/blob/90900550832bcb1845e7fd48d76874c421b16be6/src/libs/SidebarUtils.js#L148-L149

This issue is caused by this check to prevent a report that contains a recently made draft from moving to the top region  and treat it as a report without draft. 
But if a report contains draft and already placed at the top / draft region then the user view the report, this check will make the report as a report without draft. So it will jump back to the bottom region where report without draft placed.

My proposal is to only change order of reports if maximum lastMessageTimestamp is different from previous max (new message arrived) or when user navigate to other report.

```diff
.../libs/SidebarUtils.js

+ // add previous states / variables:
+ let previousState = {
+    maxLastMessageTimestamp,
+    orderedReportIDs,
+   reportIDFromRoute,
+} 
function getOrderedReportIDs(reportIDFromRoute) {
....
const filteredReports = _.filter(reports, report => ReportUtils.shouldReportBeInOptionList(report, reportIDFromRoute, isInGSDMode, currentUserLogin, iouReports, betas, policies));
+ const maxLastMessageTimestamp = lodashGet(_.max(filteredReports, r =>lodashGet(r, ['lastMessageTimestamp'], 0), 'lastMessageTimestamp', 0),  "lastMessageTimestamp", 0);
....
-        } else if (report.hasDraft && report.reportID !== reportIDFromRoute) {
+        } else if (report.hasDraft) {
            draftReportOptions.push(report);
        } else {
            recentReportOptions.push(report);
        }
    }
....
- return _.pluck(recentReportOptions, 'reportID');
+ const orderedReportIDs = _.pluck(recentReportOptions, 'reportID');

+ if ( previousState.maxLastMessageTimestamp !== maxLastMessageTimestamp 
+         || previousState.reportIDFromRoute !== reportIDFromRoute
+    ) {
+       previousState = {
+            maxLastMessageTimestamp,
+            reportIDFromRoute,
+            orderedReportId : orderedReportIDs,
+       }
+        return orderedReportIDs;
+    }

+ return previousState.orderedReportIDs;

```
This will solve this issue, and if user delete a draft message from a report, the report will not immediately moving to bottom, but it will move if user select/navigate to other report or when new message arrived.

**Comment by rawalyogendra** [comment]
### Proposal

#### Problem
The problem lies in the following code block, where the `report` is not added with other `draftReports` if it is active. It is causing the jump when user opens the chat which has draft.
https://github.com/Expensify/App/blob/e38df100d62d197689bb16888a3ac7859f157681/src/libs/SidebarUtils.js#L146-L150

#### Solution
It seems like that check was added  to stop the `report` from jumping when the user is trying to create a new message. We can simply remove `report.reportID !== reportIDFromRoute` and the current behavior will be fixed, but it will cause the chat to move when the user is creating a new message.

The solution will also break the following test, which expects the `reports` at their current position even when they have draft in them.
https://github.com/Expensify/App/blob/e38df100d62d197689bb16888a3ac7859f157681/tests/unit/LHNOrderTest.js#L145-L156
It can be fixed by adding proper expect statements
```js
expect(lodashGet(displayNames, [0, 'props', 'children'])).toBe('One, Two'); // this has `hasDraft` flag enabled so it will be on top
expect(lodashGet(displayNames, [1, 'props', 'children'])).toBe('Five, Six');
expect(lodashGet(displayNames, [2, 'props', 'children'])).toBe('Three, Four');
```

https://user-images.githubusercontent.com/7548730/196043863-03fcc801-a7d1-4a3a-b1ed-745a66f7ecc6.mp4
_NOTE: The chat jumps to the top when a draft message is created_

**Comment by mallenexpensify** [comment]
@mananjadhav Can you please review the above proposals?

**Comment by tgolen** [comment]
The proposal from @rawalyogendra is more what I was looking for, so that one is my preference.

**Comment by mananjadhav** [comment]
Sorry it took a bit longer to review. @tgolen I am inclined too with @rawalyogendra's proposal. But as @rawalyogendra also mentioned, with their proposal, it'll jump the active chat to the top, the moment we type a single letter in the composer. Are we okay with that behavior?


@bernhardoj I don't think this is an efficient solution to store these history. Are you suggesting that all the chats will eventually go and be stored in `previousReportsDrafted` if I start drafting them? When/how do we clear? + keep that outside the scope of the calling function?

@tsa321 Same as @bernhardoj's proposal, storing the previous state, that too outside the function doesn't sound very efficient to me.

**Comment by tsa321** [comment]
@mananjadhav in my proposal there is an option to passing this variable as parameter. So this variable will not be outside of the function.

> We can pass `previousState` value as parameter for `getOrderedReportIDs(reportIDFromRoute, previousState)`. So we can remove `previousState` variable in `SidebarUtils.js` and save this previous state in `SidebarLinks.js`:
> 
> ```
> let optionListItems;
> [optionListItems, this.previousState] = SidebarUtils.getOrderedReportIDs(this.props.reportIDFromRoute, this.previousState);
> ```
> 
> in:
> 
> https://github.com/Expensify/App/blob/363d98fd3a3567c102c5359ba5257c67f95c60d0/src/pages/home/sidebar/SidebarLinks.js#L95

I have updated my solution above.

and yes the more efficient solution is of course better than my proposal.

In my proposal, draft will not move up immediately after user making a draft.
And if user remove draft the chat is placed at the top region (draft section), the chat will not immediately move down. 
But it will move after user select other report or new message received or current chat is pinned.

**Comment by tsa321** [comment]
Or for much cleaner code, we can move  this `shouldReorder` logic into `SidebarLinks.js`. So we don't touch `SidebarUtils.js` file.

**Comment by bernhardoj** [comment]
@mananjadhav Thanks for the feedback. Here is my updated Proposal

### **Updated Proposal**
The same concept from my previous proposal that we need to know whether the previous report is a draft or not. By taking all considerations, we can make use of `componentDidUpdate()` inside `SidebarLinks.js` to get the previous reports.

Here is the code changes of 
_**SidebarLinks.js**_
```diff
class SidebarLinks extends React.Component {
+    constructor(props) {
+       super(props);
+        this.state = {
+            optionListItems: SidebarUtils.getOrderedReportIDs(this.props.reportIDFromRoute, {}),
+        };
+    }
    
+    componentDidUpdate(prevProps) {
+        if (!_.isEqual(prevProps.reports, this.props.reports)) {
+            this.setState({
+                optionListItems: SidebarUtils.getOrderedReportIDs(this.props.reportIDFromRoute, prevProps.reports),
+            });
+        }
+    }

    showSearchPage() {
        Navigation.navigate(ROUTES.SEARCH);
    }

    render() {
        // Wait until the personalDetails are actually loaded before displaying the LHN
        if (_.isEmpty(this.props.personalDetails)) {
            return null;
        }
-       const optionListItems = SidebarUtils.getOrderedReportIDs(this.props.reportIDFromRoute);
        return (
            <View
                accessibilityElementsHidden={this.props.isSmallScreenWidth && !this.props.isDrawerOpen}
                accessibilityLabel="List of chats"
                style={[styles.flex1, styles.h100]}
            >
                ...
                <Freeze freeze={this.props.isSmallScreenWidth && !this.props.isDrawerOpen}>
                    <LHNOptionsList
                        contentContainerStyles={[
                            styles.sidebarListContainer,
                            {paddingBottom: StyleUtils.getSafeAreaMargins(this.props.insets).marginBottom},
                        ]}
-                       data={optionListItems}
+                       data={this.state.optionListItems}
-                       focusedIndex={_.findIndex(optionListItems, (
+                       focusedIndex={_.findIndex(this.state.optionListItems, (
                            option => option.toString() === this.props.reportIDFromRoute
                        ))}
                        onSelectRow={(option) => {
                            Navigation.navigate(ROUTES.getReportRoute(option.reportID));
                            this.props.onLinkClick();
                        }}
                        shouldDisableFocusOptions={this.props.isSmallScreenWidth}
                        optionMode={this.props.priorityMode === CONST.PRIORITY_MODE.GSD ? 'compact' : 'default'}
                        onLayout={App.setSidebarLoaded}
                     />
                </Freeze>
            </View>
        );
    }
}
```

And there is not much difference in **_SidebarUtils.js_** but much cleaner imo
```diff
function getOrderedReportIDs(reportIDFromRoute, prevReports) {
    ...
    // Put all the reports into the different buckets
    for (let i = 0; i < orderedReports.length; i++) {
        const report = orderedReports[i];
+       const prevReport = prevReports[`${ONYXKEYS.COLLECTION.REPORT}${report.reportID}`];

        // If the report is pinned and we are using the option to display pinned reports on top then we need to
        // collect the pinned reports so we can sort them alphabetically once they are collected
        if (report.isPinned) {
            pinnedReportOptions.push(report);
        } else if (report.hasOutstandingIOU && !report.isIOUReportOwner) {
            iouDebtReportOptions.push(report);

        // If the active report has a draft, we do not put it in the group of draft 
+       // (unless it was previously in the group of draft) 
        // reports because we want it to maintain it's current position. Otherwise the report's position
        // jumps around in the LHN and it's kind of confusing to the user to see the LHN reorder when they start typing a comment on a report.
        } else if (
            (report.hasDraft && report.reportID !== reportIDFromRoute)
+           || (prevReport?.hasDraft && report.reportID === reportIDFromRoute)
        ) {
            draftReportOptions.push(report);
        } else {
            recentReportOptions.push(report);
        }
    }
```
The result is the same as previous video.

**Comment by bernhardoj** [comment]
@mananjadhav Thanks for the feedback. Here is my updated proposal.

### **Updated Proposal**
I still using the similar approach, which is to store the previous reports as history, but now we store it in **SidebarLinks.js**. We will update the previous report every time user clicked a different chat, and we pass this previous report to `getOrderedReportIDs()`. 

```diff
class SidebarLinks extends React.Component {
+   constructor(props) {
+       super(props);
+       this.prevReports = {}
+   }

+   componentDidUpdate(prevProps) {
+       if (prevProps.reportIDFromRoute !== this.props.reportIDFromRoute) {
+           this.prevReports = prevProps.reports;
+       }
+   }

    render() {
        // Wait until the personalDetails are actually loaded before displaying the LHN
        if (_.isEmpty(this.props.personalDetails)) {
            return null;
        }
-       const optionListItems = SidebarUtils.getOrderedReportIDs(this.props.reportIDFromRoute);
+       const optionListItems = SidebarUtils.getOrderedReportIDs(this.props.reportIDFromRoute, this.prevReports);
        ....
    }
}
```

And on **SidebarUtils.js**, do the checking if the previous report is draft or not.

```diff
-function getOrderedReportIDs(reportIDFromRoute) {
+function getOrderedReportIDs(reportIDFromRoute, prevReports) {
// Put all the reports into the different buckets
    for (let i = 0; i < orderedReports.length; i++) {
        const report = orderedReports[i];
+       const prevReport = prevReports[`${ONYXKEYS.COLLECTION.REPORT}${report.reportID}`];
        // If the report is pinned and we are using the option to display pinned reports on top then we need to
        // collect the pinned reports so we can sort them alphabetically once they are collected
        if (report.isPinned) {
            pinnedReportOptions.push(report);
        } else if (report.hasOutstandingIOU && !report.isIOUReportOwner) {
            iouDebtReportOptions.push(report);

        // If the active report has a draft, we do not put it in the group of draft
+       // (unless it was previously in the group of draft) 
        // reports because we want it to maintain it's current position. Otherwise the report's position
        // jumps around in the LHN and it's kind of confusing to the user to see the LHN reorder when they start typing a comment on a report.
        } else if (
            (report.hasDraft && report.reportID !== reportIDFromRoute)
+           || prevReport?.hasDraft
        ) {
            draftReportOptions.push(report);
        } else {
            recentReportOptions.push(report);
        }
    }
```

**Comment by deetergp** [comment]
The behavior of @rawalyogendra's proposal is how I'd expect it to work and has the advantage of being very simple.

**Comment by mananjadhav** [comment]
If that's the expected behavior then we can go ahead with @rawalyogendra's proposal?

**Comment by mallenexpensify** [comment]
Couple other issues might be related to this.  Gonna put them on hold pending this fix. 
@deetergp we OK to hire @rawalyogendra?

https://github.com/Expensify/App/issues/11667
https://github.com/Expensify/App/issues/11590

**Comment by tgolen** [comment]
> Sorry it took a bit longer to review. @tgolen I am inclined too with @rawalyogendra's proposal. But as @rawalyogendra also mentioned, with their proposal, it'll jump the active chat to the top, the moment we type a single letter in the composer. Are we okay with that behavior?

Yes, this is OK. Let's please go with this proposal.

**Comment by deetergp** [comment]
> Issue not reproducible during KI retests. (First week)

Interesting. I guess let's hold on this one? If it doesn't show in the next round of KI retests, we can probably close it.

**Comment by rawalyogendra** [comment]
@mvtglobally @deetergp I am still able to reproduce this issue in **_v1.2.19-1_**

https://user-images.githubusercontent.com/7548730/197787588-74ce4908-f7de-433f-8e43-982b5a569298.mp4

**Comment by mallenexpensify** [comment]
Per @tgolen 's [comment](https://github.com/Expensify/App/issues/11330#issuecomment-1288122295) above, I assigned @rawalyogendra .  Please apply here and Ill hire https://www.upwork.com/jobs/~011c7454f55e6c9c1f 

I am able to repro this morn on desktop-staging and web-staging (assuming I'm testing right, draft chat is moving around in LHN) 

https://user-images.githubusercontent.com/22508304/198057954-ce69899f-cc51-428d-9f99-f01d88860543.mp4

**Comment by rawalyogendra** [comment]
> Per @tgolen 's [comment](https://github.com/Expensify/App/issues/11330#issuecomment-1288122295) above, I assigned @rawalyogendra . Please apply here and Ill hire https://www.upwork.com/jobs/~011c7454f55e6c9c1f
> 
> I am able to repro this morn on desktop-staging and web-staging (assuming I'm testing right, draft chat is moving around in LHN)

@mallenexpensify yes, that is the issue we have.

**Comment by rawalyogendra** [comment]
@mallenexpensify I have applied for job in Upwork. 

I will submit a PR to fix this issue by Friday, 28th October 2022.

Thank you.

**Comment by mananjadhav** [comment]
@rawalyogendra  Any updates on this one?

**Comment by rawalyogendra** [comment]
@mananjadhav I have not been hired for this job on Upwork. I have the PR ready but I did not know how to proceed in this case.

**Comment by mananjadhav** [comment]
@mallenexpensify Can you please hire @rawalyogendra on Upwork to move forward with this?


@rawalyogendra Assigning the job on Github [here](https://github.com/Expensify/App/issues/11330#issuecomment-1292162257) is as goo as getting hired. But yeah you can sure wait to get hired on Upwork.

Once that is done, you can submit the PR along with the filled checklist.

**Comment by mallenexpensify** [comment]
My bad, I 👍 @rawalyogendra 's post above but didn't hire in Upwork :ohnothing:
I've hired @rawalyogendra and @mananjadhav in Upwork, please submit the PR

**Comment by mvtglobally** [comment]
Issue not reproducible during KI retests. (Second week)

**Comment by mallenexpensify** [comment]
@NicMendonca Reassigning as I'm heading OOO for a month

**Comment by mvtglobally** [comment]
Issue not reproducible during KI retests. (Third week) Are we ok to close this?

**Comment by mananjadhav** [comment]
@mvtglobally the PR is yet to hit production.

**Comment by mvtglobally** [comment]
Issue not reproducible during KI retests. (Fourth week)

**Comment by mananjadhav** [comment]
@NicMendonca Quick bump on the Upwork payment here.

**Comment by NicMendonca** [comment]
@mananjadhav @rawalyogendra paid, thanks!

**Comment by mananjadhav** [comment]
Thanks @NicMendonca

**Comment by rawalyogendra** [comment]
Thank you @NicMendonca

**Comment by deetergp** [comment]
Anyone know off the top of their heads which PR might have caused this regression?

**Comment by deetergp** [comment]
[This PR](https://github.com/Expensify/App/pull/11450) was the last to touch the line that @rawalyogendra's PR fixed and seems a likely culprit.

**Comment by deetergp** [comment]
Also, I don't know who needs to hear this, but `git log -p --follow -- <file whose history you want to search>` is `:chefkiss:`

**Comment by NicMendonca** [comment]
okay all set, thanks @deetergp!
